File: impacted-test-engine/src/main/java/com/teamscale/test_impacted/engine/InternalImpactedTestEngine.java
package com.teamscale.test_impacted.engine;

import com.teamscale.report.testwise.model.TestExecution;
import com.teamscale.test_impacted.commons.LoggerUtils;
import com.teamscale.test_impacted.engine.executor.AvailableTests;
import com.teamscale.test_impacted.engine.executor.ITestSorter;
import com.teamscale.test_impacted.engine.executor.TeamscaleAgentNotifier;
import com.teamscale.test_impacted.engine.executor.TestwiseCoverageCollectingExecutionListener;
import com.teamscale.test_impacted.test_descriptor.ITestDescriptorResolver;
import com.teamscale.test_impacted.test_descriptor.TestDescriptorResolverRegistry;
import com.teamscale.test_impacted.test_descriptor.TestDescriptorUtils;
import org.junit.platform.engine.EngineDiscoveryRequest;
import org.junit.platform.engine.ExecutionRequest;
import org.junit.platform.engine.TestDescriptor;
import org.junit.platform.engine.TestEngine;
import org.junit.platform.engine.UniqueId;
import org.junit.platform.engine.support.descriptor.EngineDescriptor;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.ServiceLoader;
import java.util.logging.Logger;

import static com.teamscale.test_impacted.engine.ImpactedTestEngine.ENGINE_ID;

/**
 * Test engine called internally to allow testing without needing a {@link ServiceLoader} for {@link TestEngine} setup.
 */
class InternalImpactedTestEngine {

	private static final Logger LOGGER = LoggerUtils.getLogger(InternalImpactedTestEngine.class);

	private final TestEngineRegistry testEngineRegistry;

	private final ITestSorter testSorter;

	private final TeamscaleAgentNotifier teamscaleAgentNotifier;

	private final TestDataWriter testDataWriter;

	private final String partition;

	InternalImpactedTestEngine(ImpactedTestEngineConfiguration configuration, String partition) {
		this.testEngineRegistry = configuration.testEngineRegistry;
		this.testSorter = configuration.testSorter;
		this.testDataWriter = configuration.testDataWriter;
		this.teamscaleAgentNotifier = configuration.teamscaleAgentNotifier;
		this.partition = partition;
	}

	/**
	 * Performs test discovery by aggregating the result of all {@link TestEngine}s from the {@link TestEngineRegistry}
	 * in a single engine {@link TestDescriptor}.
	 */
	TestDescriptor discover(EngineDiscoveryRequest discoveryRequest, UniqueId uniqueId) {
		EngineDescriptor engineDescriptor = new EngineDescriptor(uniqueId, "Teamscale Impacted Tests");

		LOGGER.fine(() -> "Starting test discovery for engine " + ENGINE_ID);

		for (TestEngine delegateTestEngine : testEngineRegistry) {
			LOGGER.fine(() -> "Starting test discovery for delegate engine: " + delegateTestEngine.getId());
			TestDescriptor delegateEngineDescriptor = delegateTestEngine.discover(discoveryRequest,
					UniqueId.forEngine(delegateTestEngine.getId()));

			engineDescriptor.addChild(delegateEngineDescriptor);
		}

		LOGGER.fine(() -> "Discovered test descriptor for engine " + ENGINE_ID + ":\n" + TestDescriptorUtils
				.getTestDescriptorAsString(engineDescriptor));

		return engineDescriptor;
	}

	/**
	 * Executes the request by requesting execution of the {@link TestDescriptor} children aggregated in
	 * {@link #discover(EngineDiscoveryRequest, UniqueId)} with the corresponding {@link TestEngine}.
	 */
	void execute(ExecutionRequest request) {
		TestDescriptor rootTestDescriptor = request.getRootTestDescriptor();
		AvailableTests availableTests = TestDescriptorUtils.getAvailableTests(rootTestDescriptor, partition);

		LOGGER.fine(() -> "Starting selection and sorting " + ENGINE_ID + ":\n" + TestDescriptorUtils
				.getTestDescriptorAsString(rootTestDescriptor));

		testSorter.selectAndSort(rootTestDescriptor);

		LOGGER.fine(() -> "Starting execution of request for engine " + ENGINE_ID + ":\n" + TestDescriptorUtils
				.getTestDescriptorAsString(rootTestDescriptor));

		List<TestExecution> testExecutions = executeTests(request, rootTestDescriptor);

		testDataWriter.dumpTestExecutions(testExecutions);
		testDataWriter.dumpTestDetails(availableTests.getTestList());
		teamscaleAgentNotifier.testRunEnded();
	}

	private List<TestExecution> executeTests(ExecutionRequest request, TestDescriptor rootTestDescriptor) {
		List<TestExecution> testExecutions = new ArrayList<>();
		for (TestDescriptor engineTestDescriptor : rootTestDescriptor.getChildren()) {
			Optional<String> engineId = engineTestDescriptor.getUniqueId().getEngineId();
			if (!engineId.isPresent()) {
				LOGGER.severe(
						() -> "Engine ID for test descriptor " + engineTestDescriptor + " not present. Skipping execution of the engine.");
				continue;
			}

			TestEngine testEngine = testEngineRegistry.getTestEngine(engineId.get());
			ITestDescriptorResolver testDescriptorResolver = TestDescriptorResolverRegistry
					.getTestDescriptorResolver(testEngine.getId());
			TestwiseCoverageCollectingExecutionListener executionListener =
					new TestwiseCoverageCollectingExecutionListener(teamscaleAgentNotifier,
							testDescriptorResolver,
							request.getEngineExecutionListener());

			testEngine.execute(new ExecutionRequest(engineTestDescriptor, executionListener,
					request.getConfigurationParameters()));

			testExecutions.addAll(executionListener.getTestExecutions());
		}
		return testExecutions;
	}
}

File: impacted-test-engine/src/main/kotlin/com/teamscale/test_impacted/engine/InternalImpactedTestEngine.kt
package com.teamscale.test_impacted.engine

import com.teamscale.test_impacted.commons.LoggerUtils.createLogger
import com.teamscale.test_impacted.engine.executor.TestwiseCoverageCollectingExecutionListener
import com.teamscale.test_impacted.test_descriptor.TestDescriptorResolverRegistry.getTestDescriptorResolver
import com.teamscale.test_impacted.test_descriptor.TestDescriptorUtils.getAvailableTests
import com.teamscale.test_impacted.test_descriptor.TestDescriptorUtils.getTestDescriptorAsString
import org.junit.platform.engine.EngineDiscoveryRequest
import org.junit.platform.engine.ExecutionRequest
import org.junit.platform.engine.TestDescriptor
import org.junit.platform.engine.UniqueId
import org.junit.platform.engine.support.descriptor.EngineDescriptor

/**
 * Internal test engine implementation for executing impacted tests. This class provides functionality
 * for discovering and executing tests based on their impacted status, as determined by the provided
 * configuration and partitioning logic.
 *
 * @constructor Initializes the test engine with the given configuration and partition.
 * @param configuration The configuration object that provides dependencies such as test engine registry,
 * test sorter, and test data writer.
 * @param partition The partition identifier used to divide tests and manage their execution.
 */
internal class InternalImpactedTestEngine(
	configuration: ImpactedTestEngineConfiguration,
	private val partition: String?
) {
	private val testEngineRegistry = configuration.testEngineRegistry
	private val testSorter = configuration.testSorter
	private val teamscaleAgentNotifier = configuration.teamscaleAgentNotifier
	private val testDataWriter = configuration.testDataWriter

	/**
	 * Performs test discovery by aggregating the result of all [TestEngine]s from the [TestEngineRegistry]
	 * in a single engine [TestDescriptor].
	 */
	fun discover(discoveryRequest: EngineDiscoveryRequest?, uniqueId: UniqueId?): TestDescriptor {
		val engineDescriptor = EngineDescriptor(uniqueId, "Teamscale Impacted Tests")

		LOG.fine { "Starting test discovery for engine " + ImpactedTestEngine.ENGINE_ID }

		testEngineRegistry.forEach { delegateTestEngine ->
			LOG.fine { "Starting test discovery for delegate engine: " + delegateTestEngine.id }
			val delegateEngineDescriptor = delegateTestEngine.discover(
				discoveryRequest,
				UniqueId.forEngine(delegateTestEngine.id)
			)

			engineDescriptor.addChild(delegateEngineDescriptor)
		}

		LOG.fine {
			"Discovered test descriptor for engine ${ImpactedTestEngine.ENGINE_ID}:\n${
				getTestDescriptorAsString(engineDescriptor)
			}"
		}

		return engineDescriptor
	}

	/**
	 * Executes the request by requesting execution of the [TestDescriptor] children aggregated in
	 * [.discover] with the corresponding [org.junit.platform.engine.TestEngine].
	 */
	fun execute(request: ExecutionRequest) {
		val rootTestDescriptor = request.rootTestDescriptor
		val availableTests = getAvailableTests(rootTestDescriptor, partition)

		LOG.fine {
			"Starting selection and sorting ${ImpactedTestEngine.ENGINE_ID}:\n${
				getTestDescriptorAsString(rootTestDescriptor)
			}"
		}

		testSorter.selectAndSort(rootTestDescriptor)

		LOG.fine {
			"Starting execution of request for engine ${ImpactedTestEngine.ENGINE_ID}:\n${
				getTestDescriptorAsString(rootTestDescriptor)
			}"
		}

		val testExecutions = executeTests(request, rootTestDescriptor)

		testDataWriter.dumpTestExecutions(testExecutions)
		testDataWriter.dumpTestDetails(availableTests.testList)
		teamscaleAgentNotifier.testRunEnded()
	}

	private fun executeTests(request: ExecutionRequest, rootTestDescriptor: TestDescriptor) =
		rootTestDescriptor.children.flatMap { engineTestDescriptor ->
			val engineId = engineTestDescriptor.uniqueId.engineId
			if (!engineId.isPresent) {
				LOG.severe { "Engine ID for test descriptor $engineTestDescriptor not present. Skipping execution of the engine." }
				return@flatMap emptyList()
			}

			val testEngine = testEngineRegistry.getTestEngine(engineId.get()) ?: return@flatMap emptyList()
			val testDescriptorResolver = getTestDescriptorResolver(testEngine.id) ?: return@flatMap emptyList()
			val executionListener =
				TestwiseCoverageCollectingExecutionListener(
					teamscaleAgentNotifier,
					testDescriptorResolver,
					request.engineExecutionListener
				)

			testEngine.execute(
				ExecutionRequest(
					engineTestDescriptor, executionListener,
					request.configurationParameters
				)
			)

			executionListener.testExecutions
		}

	companion object {
		private val LOG = createLogger()
	}
}
